vllm/v1/attention/backends/gdn_attn.py

```diff
@@ -241,10 +369,156 @@ def build(  # type: ignore[override]
                     dim=0,
                     out=non_spec_query_start_loc[1:],
                 )
+                query_lens_cpu = m.query_start_loc_cpu[1:] - m.query_start_loc_cpu[:-1]
+                non_spec_query_start_loc_cpu = torch.zeros(
+                    query_lens_cpu.size(0) - num_spec_decodes + 1,
+                    dtype=torch.int32,
+                )
+                torch.cumsum(
+                    query_lens_cpu[~spec_sequence_masks.cpu()],
+                    dim=0,
+                    out=non_spec_query_start_loc_cpu[1:],
+                )
 
             assert num_accepted_tokens is not None
             num_accepted_tokens = num_accepted_tokens[spec_sequence_masks]
 
+        if enable_apc:
+            block_table_tensor_full = m.block_table_tensor
+            block_size = self.kv_cache_spec.block_size
+            num_computed_tokens_device = m.num_computed_tokens_cpu.to(
+                self.device, dtype=torch.int32
+            )
+            seq_lens_device = m.seq_lens.to(self.device, dtype=torch.int32)
+
+            block_idx_last_computed_all = (
+                (cdiv(num_computed_tokens_device, block_size) - 1)
+                .clamp(min=0)
+                .to(torch.int32)
+            )
+            block_idx_first_scheduled_all = (
+                cdiv(num_computed_tokens_device + 1, block_size) - 1
+            ).to(torch.int32)
+            block_idx_last_scheduled_all = (cdiv(seq_lens_device, block_size) - 1).to(
+                torch.int32
+            )
+
+            if spec_sequence_masks is not None:
```

Comment:
If I understand correctly, this PR does not attempt to support APC + spec dec. Could we simplify this logic by just raising if spec decode is enabled?